kf_lib_data_ingest/etl/ingest_pipeline.py

```diff
@@ -151,20 +154,89 @@ def run(self):
         exception handler so that exceptions are logged.
         """
         self.logger.info('BEGIN data ingestion.')
+        self.stage_outputs = {}
+        self.stage_discovery_sources = {}
+        self.stage_discovery_links = {}
+
         # Top level exception handler
         # Catch exception, log it to file and console, and exit
         try:
             # Iterate over stages and execute them
             output = None
-            for s in self._iterate_stages():
+            for stage in self._iterate_stages():
                 if not output:  # First stage gets no input
-                    output, checks_passed, accounting_data = s.run()
+                    output, concept_discovery_dict = stage.run()
                 else:
-                    output, checks_passed, accounting_data = s.run(output)
+                    output, concept_discovery_dict = stage.run(output)
+
+                # Collapse stage subtypes to their bases for storage
+                # (e.g. GuidedTransformStage -> TransformStage)
+                stage_type = type(stage)
+                while stage_type.__base__ != IngestStage:
+                    stage_type = stage_type.__base__
+
+                write_json(
+                    concept_discovery_dict,
+                    os.path.join(
+                        self.ingest_output_dir,
+                        stage_type.__name__ + '_concept_discovery.json'
+                    )
+                )
+
+                self.stage_outputs[stage_type] = output
+                self.validate_stage_counts(
```

Comment:
Shouldn't we collect the pass/fail result of each stage and then exit 1 if any are False?

Reply:
I forget why we want exiting?

Reply:
We talked about the following scenarios: 

Running `kidsfirst ingest <pkg>` does:
```
For each ingest stage
    - Run normal ingest pipeline stage
    - Compute standard counts using stage output
    - Validate standard counts against expectations
Then run user defined data validation tests if there are any
Collect pass/fail flags from all of the above. 
If any one of them are non-zero, exit 1. 
```
We *could* add a `--exit_on_validation_fail` flag as you had suggested before (you called it 
`--strict_accounting` I think.

Running `kidsfirst test <pkg>` effectively does:
```
kidsfirst ingest <pkg> --dry_run --exit_on_validation_fail (if we implement this opt)

Same steps as above except nothing is loaded into target service
```

Reply:
Right, ok.

> We could add a --exit_on_validation_fail flag as you had suggested before (you called it

Not going to though, because loading bad data into the dataservice isn't really a problem.